mipengine/common/node_catalogue.py

```diff
@@ -13,34 +11,74 @@
 
 @dataclass_json
 @dataclass
-class Node:
+class GlobalNode:
     nodeId: str
     rabbitmqURL: str
     monetdbURL: str
-    data: Dict[str, List[str]]
+
+
+@dataclass_json
+@dataclass
+class NodePathology:
+    name: str
+    datasets: List[str]
+
+
+@dataclass_json
+@dataclass
+class NodeData:
+    pathologies: List[NodePathology]
+
+
+@dataclass_json
+@dataclass
+class LocalNode(GlobalNode):
+    data: NodeData
+
+
+@dataclass_json
+@dataclass
+class Nodes:
+    globalNode: GlobalNode
+    localNodes: List[LocalNode]
 
```

Comment:
```
 class GlobalNode: 
 class NodePathology: 
 class NodeData: 
 class LocalNode(GlobalNode):
 class Nodes: 
```
are not needed. The only 2 classes needed to express everything here, are:

```
@dataclass_json
@dataclass
class Node:
    nodeId: str
    rabbitmqURL: str
    monetdbURL: str
    data: Optional[List["Pathology"]]=None #global node will not have pathologies

@dataclass_json
@dataclass
class Pathology:
    name: str
    datasets: List[str]
```

then inside 
```
class NodeCatalogue(metaclass=Singleton): 
   _nodes: [Node]
    def __init__(self):
        ...
```
only the __init__ should be defined

```
these are also not needed
    _datasets: Dict[str, List[str]]
    _nodes_per_dataset: Dict[str, List[LocalNode]]

as well as these:
    def pathology_exists(self, pathology: str):
    def dataset_exists(self, pathology: str, dataset: str):
    def get_nodes_with_datasets(self, datasets: List[str]):
```
    
Determining the opposite relation, nodes per dataset is not the responsibility of the NodeCatalogue. The node catalogue provides only the information about the nodes. If a part of the system needs to know which nodes contain a certain dataset this part will have to figure that out, since it has all the information needed to do that. The only purpose of NodeCatalogue.py should be to transform the NodeCatalogue.json to a python object, nothing else. What if another part of the system needed to know which nodes contain a specific pathology instead of specific datasets? It would be the responsibility of that part, given the node information, to figure that out, we would not write a method in the NodeCatalogue to do that as well. Abstraction and Separation of concerns, principles.

Reply:
1. NodeCatalogue's classes wouldn't work the way you stated because:
a. A Node doesn't have a data field that is a List[Pathology] it has an object, that contains a "pathologies" field that contains a List[Pathology]

```
"data": {
    "pathologies": [
        {
            "name": "mentalhealth",
            "datasets": [
                "mentalhealth_demo"
            ]
        },
        {
            "name": "tbi",
            "datasets": [
                "tbi_demo2"
            ]
        }
    ]
} 
```

b. The _nodes are again not a List[Node] but an object with 2 fields,  "localNodes" and "globalNode", that are a List[Node].


```
{
  "globalNode": {
    .....
  },
  "localNodes": [
  ....
  ]
}
```
c. Having a single Node class and making the data Optional would work but we would lose the automatic check when loading the data. If a local node doesn't have a data field, that's a problem and it should throw an exception.
  

2. Concerning the NodeCatalogue and modularity. From wikipedia [https://en.wikipedia.org/wiki/Separation_of_concerns](url)

> 
> Modularity, and hence separation of concerns, is achieved by **encapsulating** information inside a section of code that has a well-defined interface. Encapsulation is a means of information hiding.[2] ...
> 
> Separation of concerns results in more degrees of freedom for some aspect of the program's design, deployment, or usage. Common among these is increased freedom for simplification and maintenance of code. When concerns are well-separated, there are more opportunities for module upgrade, reuse, and independent development. **Hiding the implementation details of modules behind an interface enables improving or modifying a single concern's section of code without having to know the details of other sections and without having to make corresponding changes to those other sections.** 

NodeCatalogue is not a NodeCatalogueJsonReader, it's the module that contains the node catalogue logic, that resides in a json or a kubernetes api or wherever. It holds all the information about the nodes and exposes an interface to the system, hence encapsulation.

If we need to add a method to the NodeCatalogue to find the pathologies_per_node, then we add it in the one place that is concerned with the nodes' logic, the NodeCatalogue.

What you suggest is the exact opposite of separation of concerns where we will have to write code, concerning the node catalogue in multiple places of the code.

What if the node catalogue json format changed, or we changed it into a service instead of json? We would have to go all over the code making changes.

Maybe I am wrong...

Reply:
Η αλήθεια ειναι οτι η διαφωνία μου στο να υπαρχουν μέθοδοι που επεξεργάζονται κατα οποιοδήποτε τρόπο τη πληροφορια που προέρχεται απο το node_catalogue.json, πηγάζει απο 2 πράγματα. 
1. Οτι αυτός ειναι ενας προσωρινός μηχανισμός και 
2. το οτι αυτό το μηχανισμό τον εχουμε ονομάσει NodeCatalogue και άρα δεν εχει λόγο να παρέχει functionality που εχει να κάνει με τα data που περιέχονται σε κάθε κομβο. Νεβερδελεςς, ειναι αλήθεια οτι αρχικά αυτό το μηχανισμό τον ονομάζαμε data catalogue και υπο αυτό το πρίσμα αυτα που λες ειναι σωστα, οτι δηλαδη θα πρπει να μπορείς να το ρωτησεις "ποιοι κόμβοι περιέχουν το ταδε dataset" αντι να γίνεται αυτη η επεξεργασία μεσα στον controller.
Οποτε ας το ονομάσουμε NodeDataCatalog (not catalogue), το υπόλοιπο το κρατάμε ετσι όπως το χεις και προσθεσε 2 ακόμα μέθοδους, get_global node() και μια get_local_nodes()

Reply:
Οκ, θα προσθέσω τις μεθόδους.

Σίγουρα να το πούμε NodeDataCatalog; Δεν θα περιέχει πληροφορίες μόνο για τα ντάτα των κόμβων αλλά και για τους ίδιους τους κόμβους π.χ. rabbitmqURL, monetdbURL.